Include/internal/pycore_backoff.h

```diff
@@ -0,0 +1,117 @@
+
+#ifndef Py_INTERNAL_BACKOFF_H
+#define Py_INTERNAL_BACKOFF_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+/* 16-bit countdown counters using exponential backoff.
+
+   These are used by the adaptive specializer to count down until
+   it is time to specialize an instruction. If specialization fails
+   the counter is reset using exponential backoff.
+
+   Another use is for the Tier 2 optimizer to decide when to create
+   a new Tier 2 trace (executor). Again, exponential backoff is used.
+
+   The 16-bit counter is structured as a 12-bit unsigned 'value'
+   and a 4-bit 'backoff' field. When resetting the counter, the
+   backoff field is incremented (until it reaches a limit) and the
+   value is set to a bit mask representing the value 2**backoff - 1.
+   The maximum backoff is 12 (the number of value bits).
+
+   There is an exceptional value which must not be updated, 0xFFFF.
+*/
+
+typedef struct {
+    union {
+        uint16_t counter;
```

Comment:
Is looks like this is just used for `is_unreachable_backoff_counter`. Maybe have `backoff == 15 && value == MAX` indicate that it is unreachable?

Reply:
It's also used to get the full 16-bit value as an int. And I like that the unreachable value cannot be constructed from a (value, backoff) pair because the backoff needs to be <= 12.

Reply:
(I had to break down and remove the `assert(backoff <= 12)` because there were corner cases where the "unreachable" value would be put back in line this. But I still feel that `.counter` is useful -- it is how a `backoff_counter_t` value is converted to a plain `uint16_t` to store in certain (more) public spots.